README.md
# Shippo Node.js API wrapper
[![npm version](https://badge.fury.io/js/shippo.svg)](https://badge.fury.io/js/shippo)
[![Build Status](https://travis-ci.org/goshippo/shippo-node-client.svg?branch=add-travis-ci)](https://travis-ci.org/goshippo/shippo-node-client)

Shippo is a shipping API that connects you with [multiple shipping carriers](https://goshippo.com/carriers/) (such as USPS, UPS, DHL, Canada Post, Australia Post, UberRUSH and many others) through one interface.

Print a shipping label in 10 mins using our default USPS and DHL Express accounts. No need to register for a carrier account to get started.

You will need to [register for a Shippo account](https://goshippo.com/) to use the Shippo API. It's free to sign up, free to use the API. Only pay to print a live label, test labels are free.

## Installation:
You can install this package by running the following command:
```shell
  npm install shippo
```
This means, you don't actually have to download this repository. If you wish to make modifications to the wrapper, you can clone this repository into your project.

### Requirements:
The shippo Node.js has no additional dependencies.

## Usage:

Initialize your `shippo` instance using your `Private Auth Token` provided to you on the `API` page in the Shippo Dashboard.

```js
    var shippo = require('shippo')('<YOUR_PRIVATE_KEY>');
```
The snippet below demonstrates how to create an address object (a Shippo `Resource`). Check examples.js for more detailed example for generating a shipping label:

```js
    var shippo = require('shippo')('<YOUR_PRIVATE_KEY>');

    shippo.address.create({
          'name' : 'Mr Hippo',
          'company' : 'SF Zoo',
          'street1' : '2945 Sloat Blvd',
          'city' : 'San Francisco',
          'state' : 'CA',
          'zip' : '94132',
          'country' : 'US',
          'phone' : '+1 555 341 9393',
          'email' : 'mrhippo@goshippo.com'
    }).then(function(address){
      console.log("shipment : %s", JSON.stringify(address));
    });
```

### International Multipiece Shipment Example

This example demonstrates how to purchase a label for an international shipment.
Creating domestic shipment would follow a similiar proccess but would not require
the creation of CustomsItems and CustomsDeclaration objects.

```js

var addressFrom  = {
	"name":"Ms Hippo",
	"company":"Shippo",
	"street1":"215 Clayton St.",
	"city":"San Francisco",
	"state":"CA",
	"zip":"94117",
	"country":"US", //iso2 country code
	"phone":"+1 555 341 9393",
	"email":"support@goshippo.com",
};

// example address_to object dict
var addressTo = {
	"name":"Ms Hippo",
	"company":"Shippo",
	"street1":"803 Clayton St.",
	"city":"San Francisco",
	"state":"CA",
	"zip":"94117",
	"country":"US", //iso2 country code
	"phone":"+1 555 341 9393",
	"email":"support@goshippo.com",
};

// parcel object dict
var parcelOne = {
	"length":"5",
	"width":"5",
	"height":"5",
	"distance_unit":"in",
	"weight":"2",
	"mass_unit":"lb"
};

var parcelTwo = {
    "length":"5",
    "width":"5",
    "height":"5",
    "distance_unit":"in",
    "weight":"2",
    "mass_unit":"lb"
};

var shipment = {
    "address_from": addressFrom,
    "address_to": addressTo,
    "parcels": [parcelOne, parcelTwo],
};

shippo.transaction.create({
	"shipment": shipment,
	"servicelevel_token": "ups_ground",
	"carrier_account": "558c84bbc25a4f609f9ba02da9791fe4",
	"label_file_type": "png"
})
.then(function(transaction) {
    shippo.transaction.list({
      "rate": transaction.rate
    })
    .then(function(mpsTransactions) {
        mpsTransactions.results.forEach(function(mpsTransaction){
            if(mpsTransaction.status == "SUCCESS") {
                console.log("Label URL: %s", mpsTransaction.label_url);
                console.log("Tracking Number: %s", mpsTransaction.tracking_number);
            } else {
                // hanlde error transactions
                console.log("Message: %s", mpsTransactions.messages);
            }
        });
    })
}, function(err) {
    // Deal with an error
    console.log("There was an error creating transaction : %s", err.detail);
});
```

## Tests:
### Requirements:
For the test cases the following packages are required:
```js
  npm install mocha
  npm install chai
  npm install mocha-as-promised
  npm install chai-as-promised
```

## Credits

This project was influenced by the excellent [stripe-node](https://github.com/stripe/stripe-node).

## Documentation

Please see [https://goshippo.com/docs](https://goshippo.com/docs) for up-to-date documentation.

## About Shippo

Connect with multiple different carriers, get discounted shipping labels, track parcels, and much more with just one integration. You can use your own carrier accounts or take advantage of our discounted rates with the USPS and DHL Express. Using Shippo makes it easy to deal with multiple carrier integrations, rate shopping, tracking and other parts of the shipping workflow. We provide the API and dashboard for all your shipping needs.

## Supported Features

The Shippo API provides in depth support of carrier and shipping functionalities. Here are just some of the features we support through the API:

* Shipping rates & labels
* Tracking for any shipment with just the tracking number
* Batch label generation
* Multi-piece shipments
* Manifests and SCAN forms
* Customs declaration and commercial invoicing
* Address verification
* Signature and adult signature confirmation
* Consolidator support including:
	* DHL eCommerce
	* UPS Mail Innovations
	* FedEx Smartpost
* Additional services: cash-on-delivery, certified mail, delivery confirmation, and more.


CHANGELOG
1.7.1 Feb 2nd, 2022
- add better error handling

1.7.0 Jan 18, 2022
- add pickup + example

1.6.0 Jan 20, 2021
- included additional fields for CustomsItems and CustomsDeclarations in the tests and example

1.3.0 March 30, 2017
- add support for new API version
- for a complete list of API changes see https://goshippo.com/docs/changelog

1.2.0 Jan 18, 2017
 - add batch and tracking + examples

1.1.2 Jun 3rd, 2016
 - add Travis CI and badges

1.1.1 Jun 19th, 2015
 - added support for CarrierAccounts
 - changed authentication to use ShippoToken (from http basic auth)

1.0.0 Aug 25th, 2014
 - changed class names to singular
 - Added extensive unit tests
 - added optional parameter <currency> for shipment.rates()
 - added optional parameters <size> and <page> for list method
 - added validate method to address
 - added manifest, rate, and refund classes
 - added “s” to parcel class
 - made fix to retrieve method 
 
0.0.1 Mar 21th, 2014
 - Initial version, HEAVILY inspired by the excellent stripe-node 


example.js
/**
This example demonstrates how to purchase a label for an international shipment.
Creating domestic shipment would follow a similiar proccess but would not require
the creation of CustomsItems and CustomsDeclaration objects.
**/


// replace <YOUR_PRIVATE_KEY> with your ShippoToken key
var shippo = require('shippo')('<YOUR_PRIVATE_KEY>');

var addressFrom  = {
	"name":"Ms Hippo",
	"company":"Shippo",
	"street1":"215 Clayton St.",
	"city":"San Francisco",
	"state":"CA",
	"zip":"94117",
	"country":"US", //iso2 country code
	"phone":"+1 555 341 9393",
	"email":"ms-hippo@goshippo.com",
}

// example address_to object dict
var addressTo = {
	"name":"Mr Hippo",
	"company":"London Zoo",
	"street1":"Regent's Park",
	"street2":"Outer Cir",
	"city":"LONDON",
	"state":"",
	"zip":"NW1 4RY",
	"country":"GB", //iso2 country code
	"phone":"+1 555 341 9393",
	"email":"mrhippo@goshippo.com",
	"metadata" : "Hippo T-Shirt Order #1043"
}

// parcel object dict
var parcel = {
	"length":"5",
	"width":"5",
	"height":"5",
	"distance_unit":"in",
	"weight":"2",
	"mass_unit":"lb",
}

// example CustomsItems object. This is required for int'l shipment only.
var customsItem = {
	"description":"T-Shirt",
	"quantity":2,
	"net_weight":"0.3",
	"mass_unit":"lb",
	"value_amount":"20",
	"value_currency":"USD",
	"origin_country":"US",
	"eccn_ear99": "3A001"
}

var carrier_account;
let transactionCanBePickedUp = false;
// Creating the CustomsDeclaration
// (CustomsDeclaration are NOT required for domestic shipments)
shippo.customsdeclaration.create({
	"certify_signer": "Mr. Hippo",
	"certify": true,
	"items": [customsItem],
	"non_delivery_option": "RETURN",
	"contents_type": "MERCHANDISE",
	"contents_explanation": "stuff",
	"exporter_reference": "Exporter Reference",
	"importer_reference": "",
	"invoice": "Invoice",
	"license": "License",
	"certificate": "Certificate",
	"notes": "Notes",
	"eel_pfc": "NOEEI_30_37_a",
	"aes_itn": "X20180426506889",
	"incoterm": "DDP",
	"b13a_filing_option": "FILED_ELECTRONICALLY",
	"b13a_number": "AB1234567890123",
	"invoiced_charges": {
		"currency": "USD",
		"total_shipping": "1.23",
		"total_taxes": "4.56",
		"total_duties": "78.90",
		"other_fees": "9.87"
	},
	"metadata": "test customsDeclaration"
}).catch(function(err) {
	// Deal with an error
	if (err instanceof shippo.error.ShippoAuthenticationError) {
		console.log("There was an error authenticating against the Shippo API");
	} else {
		console.log("There was an error creating customs declaration: %s", err);
	}
	process.exit(1);
}).then(function(customsDeclaration) {
	console.log("customs Declaration : %s", JSON.stringify(customsDeclaration, null, 4));
	// Creating the shipment object. In this example, the objects are directly passed to the
	// shipment.create method, Alternatively, the Address and Parcel objects could be created
	// using address.create(..) and parcel.create(..) functions respectively.
	// adding the async:false makes this call synchronous
	return shippo.shipment.create({
		"address_from": addressFrom,
		"address_to": addressTo,
		"parcels": [parcel],
		"customs_declaration": customsDeclaration.object_id,
		"async": false
	});
}).catch(function(err) {
	// Deal with an error
	console.log("There was an error creating shipment: %s", err);
	process.exit(1);
}).then(function(shipment) {
	console.log("shipment : %s", JSON.stringify(shipment, null, 4));
	return shippo.shipment.rates(shipment.object_id);
}).catch(function(err) {
	// Deal with an error
	console.log("There was an error retrieving rates : %s", err);
	process.exit(1);
}).then(function(rates) {
	console.log("rates : %s", JSON.stringify(rates, null, 4));
	try {
		const filteredRates = rates.results.filter(rate => rate.provider.toUpperCase().includes('USPS') || rate.provider.toUpperCase().includes('DHL EXPRESS'));
		transactionCanBePickedUp = true;
		rate = filteredRates[0];
		carrier_account = rate.carrier_account;
	} catch (err) {
		// Get the first rate in the rates results for demo purposes.
		rate = rates.results[0];
		if (rates.results[0]) {
			carrier_account = rates.results[0].carrier_account;
		}
	}
	// Purchase the desired rate
	return shippo.transaction.create({"rate": rate.object_id, "async": false})
}).catch(function(err) {
	// Deal with an error
	console.log("There was an error creating transaction : %s", err);
	process.exit(1);
}).then(function(transaction) {
	console.log("transaction : %s", JSON.stringify(transaction, null, 4));
	// print label_url and tracking_number
	if(transaction.status == "SUCCESS") {
		console.log("Label URL: %s", transaction.label_url);
		console.log("Tracking Number: %s", transaction.tracking_number);
	}else{
		//Deal with an error with the transaction
		console.log("Message: %s", JSON.stringify(transaction.messages, null, 2));
	}
	if (transactionCanBePickedUp) {			
		var requested_start_time = new Date();
		requested_start_time.setDate(requested_start_time.getDate() + 1);
		var requested_end_time = new Date();
		requested_end_time.setDate(requested_start_time.getDate() + 1.1);
		return shippo.pickup.create({
			carrier_account,
			location: {
				building_location_type: 'Knock on Door',
				address: addressFrom,
			},
			transactions: [transaction.object_id],
			requested_start_time,
			requested_end_time,
			is_test: false
		})
	}
	return false;
}).catch(function(err) {
	if (err.detail && err.detail.messages) {
		err.detail.messages.forEach(message => {
			if(message.includes('You have already requested')) {
				console.log("There is a pickup already scheduled for this address and date");
				process.exit(1);
			}
		});		
	}
	// Deal with an error
	console.log("There was an error creating a pickup : %s", err);
	process.exit(1);
}).then(function(pickups) {
	if (transactionCanBePickedUp) {
		console.log("pickups : %s", JSON.stringify(pickups, null, 4));
	}
});
